src/backend/columnar/columnar_reader.c

```diff
@@ -297,78 +276,153 @@ ColumnarReadRowByRowNumber(Relation relation, ColumnarRandomReadState *randomRea
 		return false;
 	}
 
-	StripeMetadata *cachedStripeMetadata = randomReadState->cachedStripeMetadata;
-	if (!cachedStripeMetadata || cachedStripeMetadata->id != stripeMetadata->id)
+	if (!ColumnarReadIsCurrentStripe(readState, stripeMetadata))
 	{
 		/* do the cleanup before starting to read a new stripe */
-		ColumnarResetRandomRead(randomReadState);
+		ColumnarResetRead(readState);
 
 		TupleDesc relationTupleDesc = RelationGetDescr(relation);
 		List *whereClauseList = NIL;
 		List *whereClauseVars = NIL;
-		MemoryContext stripeReadContext = randomReadState->stripeReadContext;
-		StripeReadState *stripeReadState = BeginStripeRead(stripeMetadata,
-														   relation,
-														   relationTupleDesc,
-														   neededColumnList,
-														   whereClauseList,
-														   whereClauseVars,
-														   stripeReadContext);
-		ColumnarCacheRandomRead(randomReadState, stripeMetadata, stripeReadState);
+		MemoryContext stripeReadContext = readState->stripeReadContext;
+		readState->stripeReadState = BeginStripeRead(stripeMetadata,
+													 relation,
+													 relationTupleDesc,
+													 readState->projectedColumnList,
+													 whereClauseList,
+													 whereClauseVars,
+													 stripeReadContext);
+
+		/*
+		 * XXX: Since stripes in stripeList are ordered by their first row
+		 * numbers, we can't just assume that stripe with "id" will be the
+		 * "id"th item of given stripe list.
+		 * For this reason, we should either store StripeMetadata instead of
+		 * currentStripe, or we need to linearly search our stripe in stripe
+		 * list to set currentStripe, as in below.
+		 */
+		readState->currentStripe = StripeListGetStripeIndex(readState->stripeList,
+															stripeMetadata);
+		Assert(readState->currentStripe >= 0);
```

Comment:
Regarding https://github.com/citusdata/citus/pull/5058#discussion_r655918433, I made some changes in 01595a14515ecab4337fa0724f8767ef06fb8dc7. 

On the other hand, above part seems a bit annoying to me.
Now we don't store a redundant `StripeMetadata` object and directly make use of `readState->stripeList[readState->currentStripe]`. 
For this reason, we need to linearly search `stripeList` to correctly set `currentStripe`.

Does that look good ?
Or, should we have another field in `ColumnarReadState` to store `StripeMetadata` itself additionally ?
The second approach would make `ColumnarReadState` slightly confusing, but current approach is not very nice from the performance aspect.

What do you think @jeff-davis ?

Reply:
A list is conceptually the wrong structure -- an array would be better. But since PG13, the `List` structure has become more similar to a C++ Vector (growable array), so it's not terribly important. If you think there will be a real performance concern in PG12 we can either change it to an array or keep an extra pointer around to the current cell.

Reply:
Indeed, regardless of underlying list implementation, here we search `StripeMetadata` in given `stripeList` to correctly set `currentStripe`.
For this reason, whether we are on PG12 or PG13, `StripeListGetStripeIndex` is always O(n)`, where `n = length(stripeList)`.

Reply:
Now we don't have the `stripeList` field anymore, so resolving